### PR TITLE
SDK support for StableTokenRegistry

### LINKED ISSUE #9581: Create ContractKit for StableTokenRegistry.sol
### Expected Behavior

There should be a contract kit for StableTokenRegistry.sol

### Current Behavior

Currently no contract kit created

### PR DESCRIPTION
**Pull request by ninabarbakadze**
### Description

providing sdk support for on-chain stable token registry

### Related issues

- Fixes #9581

### CODE REVIEW

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.t.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { Address } from '@celo/base/lib/address'
```
Comment (martinvol) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.t.ts]: Forgot the test in the name of the file?

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  /**
+   * Returns the contatenated contracts and each of their lengths
+   * @return concatenatedContract, ContractLengths
+   */
+  async getContractInstances(): Promise<ContractInstance> {
+    const ret = await Promise.resolve(this.contract.methods.getContractInstances())
+    return {
+      concatenatedContracts: Object.keys(ret)[0],
+      contractLengths: Object.keys(ret)[1],
```
Comment (martinvol) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts:23]: I don't think it's needed to return the length here, as in TS you can just query the size

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  stableTokens = proxyCall(this.contract.methods.stableTokens)
+
+  /**
+   * Returns the contatenated contracts and each of their lengths
+   * @return concatenatedContract, ContractLengths
+   */
+  async getContractInstances(): Promise<ContractInstance> {
+    const ret = await Promise.resolve(this.contract.methods.getContractInstances())
+    return {
+      concatenatedContracts: Object.keys(ret)[0],
```
Comment (martinvol) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: Here you could do the formatting, directly, so that the user just gets a nice array of strings

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.t.ts (hunk 1 above)
Comment (ninabarbakadze) [reply packages/sdk/contractkit/src/wrappers/StableTokenRegistry.t.ts]: yea a typo

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.test.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+  beforeAll(async () => {
+    accounts = await web3.eth.getAccounts()
+    kit.defaultAccount = accounts[0]
+    stableTokenRegistry = await kit.contracts.getStableTokenRegistry()
+  })
+
+  describe('Verifying that it always has correct values', () => {
+    it('has the correct list of fiatTicker', () => {
+      expect(stableTokenRegistry.FiatTickers()).toEqual(['USD', 'EUR', 'BRL'])
```
Comment (martinvol) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.test.ts]: careful with the capitalization

Comment (ninabarbakadze) [reply packages/sdk/contractkit/src/wrappers/StableTokenRegistry.test.ts]: values I've hardcoded in the contract are also capitalised, but I think it's worth writing a note about it

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts (hunk 3 above)
Comment (ninabarbakadze) [reply packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: ✅

File: packages/sdk/contractkit/src/contract-cache.ts [hunk 5]
```diff
@@ -66,6 +68,7 @@ const WrapperFactoriesWhichNeedCache = {
   [CeloContract.Governance]: GovernanceWrapper,
   [CeloContract.LockedGold]: LockedGoldWrapper,
   [CeloContract.Validators]: ValidatorsWrapper,
+  [CeloContract.StableTokenRegistry]: StableTokenRegistryWrapper,
```
Comment (pedro-clabs) [inline packages/sdk/contractkit/src/contract-cache.ts]: Nit: keep alphabetical order.

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import { StableTokenRegistry } from '../generated/StableTokenRegistry'
+import { BaseWrapper, proxyCall } from './BaseWrapper'
+
+/**
+ * Returns the contatenated contracts and each of their lengths
+ * @param contractHex concatenated contracts
+ * @param lengths their lengths
+ * @return string array containing contract names
+ */
+const returnSTContractNames = (contractsHex: string, lengths: string[]): string[] => {
```
Comment (pedro-clabs) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: May I suggest renaming it to something like "splitContractNames"? Or even just something like "splitByLengths", as the input could be any arbitrary string.

**Review by pedro-clabs — COMMENTED** [review]
Thanks, Nina!

File: packages/sdk/contractkit/src/contract-cache.ts (hunk 5 above)
Comment (ninabarbakadze) [reply packages/sdk/contractkit/src/contract-cache.ts]: I think I should just remove this contract from cache section @pedro-clabs

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import Web3 from 'web3'
+import { StableTokenRegistry } from '../generated/StableTokenRegistry'
+import { BaseWrapper, proxyCall } from './BaseWrapper'
+
+/**
+ * Returns the contatenated contracts and each of their lengths
+ * @param contractHex concatenated contracts
+ * @param lengths their lengths
+ * @return string array containing contract names
```
Comment (pedro-clabs) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]:
```suggestion
 * Splits a string with concatenated contract names into each individual name.
 * @param contractHex concatenated contract names
 * @param lengths their lengths
 * @return string array containing contract names
```

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  }
+  return contractsArr
+}
+
+export class StableTokenRegistryWrapper extends BaseWrapper<StableTokenRegistry> {
+  fiatTickers = proxyCall(this.contract.methods.fiatTickers)
+  stableTokens = proxyCall(this.contract.methods.stableTokens)
+
+  /**
+   * Returns the contatenated contracts and each of their lengths
```
Comment (pedro-clabs) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: Does it? :)

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  async FiatTickers(): Promise<string[]> {
+    const convertedToHex = []
+    try {
+      let index = 0
+      while (await this.fiatTickers(index)) {
+        convertedToHex.push(Web3.utils.hexToUtf8(await this.fiatTickers(index)))
+        index++
+      }
+    } catch (error) {
+      return convertedToHex
```
Comment (pedro-clabs) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: Does it make sense to return a potentially partially filled array here? Without much context, I'd suggest rejecting the promise (e.g. `return Promise.reject(error);`). WDYT?

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    const contractLengths = values[1]
+    return splitContractNamesByLength(concatenatedContracts, contractLengths)
+  }
+
+  async FiatTickers(): Promise<string[]> {
+    const convertedToHex = []
+    try {
+      let index = 0
+      while (await this.fiatTickers(index)) {
+        convertedToHex.push(Web3.utils.hexToUtf8(await this.fiatTickers(index)))
```
Comment (pedro-clabs) [inline packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: Please correct me if I misunderstood it, but I think `fiatTickers` is being called twice here: once to check that the index is valid, and once more to actually get the value. Would it be possible to call it once and test on the return value?

Naively, I'd also suggest that we issue the calls concurrently.

Comment (ninabarbakadze) [reply packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: fiatTickers is a storage variable(an array), in the while statement, I'm just checking that a value exists on a given index.

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts (hunk 9 above)
Comment (ninabarbakadze) [reply packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: I agree

File: packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts (hunk 8 above)
Comment (ninabarbakadze) [reply packages/sdk/contractkit/src/wrappers/StableTokenRegistry.ts]: good catch, absolutely not 😁

File: packages/protocol/contracts/stability/StableTokenRegistry.sol [hunk 11]
```diff
@@ -1,9 +1,8 @@
 pragma solidity ^0.5.13;
 
-import "openzeppelin-solidity/contracts/math/SafeMath.sol";
-import "../common/Initializable.sol";
-import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
-import "../common/interfaces/IRegistry.sol";
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import { Initializable } from "../common/Initializable.sol";
+import { Ownable } from "openzeppelin-solidity/contracts/ownership/Ownable.sol";
```
Comment (pedro-clabs) [inline packages/protocol/contracts/stability/StableTokenRegistry.sol]: Is this change related to the SDK changes?

**Review by pedro-clabs — APPROVED** [review]
LGTM. Thanks, Nina!

File: packages/protocol/contracts/stability/StableTokenRegistry.sol (hunk 11 above)
Comment (ninabarbakadze) [reply packages/protocol/contracts/stability/StableTokenRegistry.sol]: no, this is only for readability

Comment (ninabarbakadze) [reply packages/protocol/contracts/stability/StableTokenRegistry.sol]: and IRegistry is not used in the contract

Comment (pedro-clabs) [reply packages/protocol/contracts/stability/StableTokenRegistry.sol]: Got it, thanks.
